File: api/internal/model/inbox.go
package model

import (
	"time"

	"github.com/google/uuid"
)

// MaxInboxItems is the hard limit on how many items a user can have in their inbox.
const MaxInboxItems = 100

// InboxPositionGap is the gap between consecutive inbox item positions.
const InboxPositionGap = 1000

// InboxItem represents a work item in a user's personal inbox.
type InboxItem struct {
	ID         uuid.UUID `json:"id"`
	UserID     uuid.UUID `json:"user_id"`
	WorkItemID uuid.UUID `json:"work_item_id"`
	Position   int       `json:"position"`
	CreatedAt  time.Time `json:"created_at"`
}

// InboxItemWithWorkItem is an InboxItem enriched with joined work item and project data.
type InboxItemWithWorkItem struct {
	InboxItem
	DisplayID      string     `json:"display_id"`
	Title          string     `json:"title"`
	Type           string     `json:"type"`
	Status         string     `json:"status"`
	StatusCategory string     `json:"status_category"`
	Priority       string     `json:"priority"`
	ProjectKey     string     `json:"project_key"`
	ProjectName    string     `json:"project_name"`
	AssigneeID          *uuid.UUID `json:"assignee_id,omitempty"`
	AssigneeDisplayName string     `json:"assignee_display_name,omitempty"`
	Description         string     `json:"description,omitempty"`
	DueDate             *time.Time `json:"due_date,omitempty"`
	SLATargetAt         *time.Time `json:"sla_target_at,omitempty"`
	UpdatedAt           time.Time  `json:"updated_at"`
}

// InboxItemList is the paginated result for inbox item listings.
type InboxItemList struct {
	Items   []InboxItemWithWorkItem `json:"items"`
	Cursor  string                 `json:"cursor"`
	HasMore bool                   `json:"has_more"`
	Total   int                    `json:"total"`
}

File: web/src/api/inbox.ts
import { api } from './client'
import type { SLAInfo } from './workitems'

// --- Types ---

export interface InboxItem {
  id: string
  work_item_id: string
  position: number
  created_at: string
  display_id: string
  title: string
  type: string
  status: string
  status_category: string
  priority: string
  project_key: string
  project_name: string
  namespace_slug: string
  namespace_name: string
  assignee_id: string | null
  assignee_display_name: string
  description: string
  due_date: string | null
  sla: SLAInfo | null
  sla_target_at: string | null
  updated_at: string
}

export interface InboxListResponse {
  items: InboxItem[]
  cursor: string
  has_more: boolean
  total: number
}

export interface InboxFilter {
  search?: string
  include_completed?: boolean
  work_item_id?: string
  project?: string[]
  cursor?: string
  limit?: number
}

// --- API Functions ---

export async function listInboxItems(filter: InboxFilter = {}): Promise<InboxListResponse> {
  const params: Record<string, string> = {}
  if (filter.search) params.search = filter.search
  if (filter.include_completed) params.include_completed = 'true'
  if (filter.work_item_id) params.work_item_id = filter.work_item_id
  if (filter.project?.length) params.project = filter.project.join(',')
  if (filter.cursor) params.cursor = filter.cursor
  if (filter.limit) params.limit = String(filter.limit)

  const { data } = await api.get('/user/inbox', { params })
  return data.data
}

export async function addToInbox(workItemId: string): Promise<void> {
  await api.post('/user/inbox', { work_item_id: workItemId })
}

export async function removeFromInbox(inboxItemId: string): Promise<void> {
  await api.delete(`/user/inbox/${inboxItemId}`)
}

export async function reorderInboxItem(inboxItemId: string, position: number): Promise<void> {
  await api.patch(`/user/inbox/${inboxItemId}`, { position })
}

export async function clearCompletedInboxItems(): Promise<{ removed: number }> {
  const { data } = await api.delete('/user/inbox/completed')
  return data.data
}

export async function getInboxCount(): Promise<number> {
  const { data } = await api.get('/user/inbox/count')
  return data.data.count
}
